client/retry.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2023 CloudWeGo Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+
+	"github.com/cloudwego/kitex/client"
+	"github.com/cloudwego/kitex/pkg/klog"
+	"github.com/cloudwego/kitex/pkg/retry"
+	"github.com/kitex-contrib/config-etcd/etcd"
+	"github.com/kitex-contrib/config-etcd/utils"
+)
+
+// WithRetryPolicy sets the retry policy from etcd configuration center.
+func WithRetryPolicy(dest, src string, etcdClient etcd.Client, uniqueID int64, opts utils.Options) []client.Option {
+	param, err := etcdClient.ClientConfigParam(&etcd.ConfigParamConfig{
+		Category:          retryConfigName,
+		ServerServiceName: dest,
+		ClientServiceName: src,
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	for _, f := range opts.EtcdCustomFunctions {
+		f(&param)
+	}
+	key := param.Prefix + "/" + param.Path
+	rc := initRetryContainer(key, dest, etcdClient, uniqueID)
+	return []client.Option{
+		client.WithRetryContainer(rc),
+		client.WithCloseCallbacks(func() error {
+			// cancel the configuration listener when client is closed.
+			etcdClient.DeregisterConfig(key, uniqueID)
+			return nil
+		}),
+		client.WithCloseCallbacks(rc.Close),
+	}
+}
+
+func initRetryContainer(key, dest string,
+	etcdClient etcd.Client, uniqueID int64,
+) *retry.Container {
+	retryContainer := retry.NewRetryContainerWithPercentageLimit()
+
+	ts := utils.ThreadSafeSet{}
+
+	onChangeCallback := func(data string, parser etcd.ConfigParser) {
+		if data == "" {
+			klog.Debugf("[etcd] %s client etcd retry: get config failed: empty config, skip...", key)
```

Comment:
We should distinguish the following cases:
- Empty string or the key does not exist: it means the user removed the config, we should update the container with an empty config (remove all existing config)
- Unable to get the config: we should print a warning log and skip this time.

Reply:
Same for other configs.

Reply:
Done , now when initialize the config and found it is empty , the callback data is `config_empty`

Reply:
This solution is not elegant: it depends on a magic value, and is not easy to understand.

Besides, when the key does not exist on startup, `callback("", parser)` is called, and  it can not be decoded correctly.

Reply:
A better solution is to callback with both the event type and config value (e.g. `{PUT, "{}"}`, `{DELTE, ""}` so that we can react properly in the callback function.

Reply:
Done,now we can react due to the eventType

Reply:
Have you tested with DELETE? It doesn't seem to work.

I think it's better we define the callback functions as a standalone function so that we can cover the branches with unit tests.